src/storage/v3/shard_rsm.cpp

```diff
@@ -0,0 +1,439 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#include <iterator>
+#include <utility>
+
+#include "storage/v3/shard_rsm.hpp"
+#include "storage/v3/vertex_accessor.hpp"
+
+using memgraph::msgs::Label;
+using memgraph::msgs::PropertyId;
+using memgraph::msgs::Value;
+using memgraph::msgs::VertexId;
+
+namespace {
+// TODO(gvolfing use come algorithm instead of explicit for loops)
+memgraph::storage::v3::PropertyValue ToPropertyValue(Value &&value) {
+  using PV = memgraph::storage::v3::PropertyValue;
+  PV ret;
+  switch (value.type) {
+    case Value::Type::Null:
+      return PV{};
+    case Value::Type::Bool:
+      return PV(value.bool_v);
+    case Value::Type::Int64:
+      return PV(static_cast<int64_t>(value.int_v));
+    case Value::Type::Double:
+      return PV(value.double_v);
+    case Value::Type::String:
+      return PV(value.string_v);
+    case Value::Type::List: {
+      std::vector<PV> list;
+      for (auto &elem : value.list_v) {
+        list.emplace_back(ToPropertyValue(std::move(elem)));
+      }
+      return PV(list);
+    }
+    case Value::Type::Map: {
+      std::map<std::string, PV> map;
+      for (auto &[key, value] : value.map_v) {
+        map.emplace(std::make_pair(key, ToPropertyValue(std::move(value))));
+      }
+      return PV(map);
+    }
+    // These are not PropertyValues
+    case Value::Type::Vertex:
+    case Value::Type::Edge:
+    case Value::Type::Path:
+      MG_ASSERT(false, "Not PropertyValue");
+  }
+  return ret;
+}
+
+Value ToValue(const memgraph::storage::v3::PropertyValue &pv) {
+  using memgraph::storage::v3::PropertyValue;
+
+  switch (pv.type()) {
+    case PropertyValue::Type::Bool:
+      return Value(pv.ValueBool());
+    case PropertyValue::Type::Double:
+      return Value(pv.ValueDouble());
+    case PropertyValue::Type::Int:
+      return Value(pv.ValueInt());
+    case PropertyValue::Type::List: {
+      std::vector<Value> list(pv.ValueList().size());
+      for (const auto &elem : pv.ValueList()) {
+        list.emplace_back(ToValue(elem));
+      }
+
+      return Value(list);
+    }
+    case PropertyValue::Type::Map: {
+      std::map<std::string, Value> map;
+      for (const auto &[key, val] : pv.ValueMap()) {
+        // maybe use std::make_pair once the && issue is resolved.
+        map.emplace(key, ToValue(val));
+      }
+
+      return Value(map);
+    }
+    case PropertyValue::Type::Null:
+      return Value{};
+    case PropertyValue::Type::String:
+      return Value(pv.ValueString());
+    case PropertyValue::Type::TemporalData: {
+      // TBD -> we need to specify this in the messages, not a priority.
+      MG_ASSERT(false, "Temporal datatypes are not yet implemented on Value!");
+      return Value{};
+    }
+  }
+}
+
+std::vector<std::pair<memgraph::storage::v3::PropertyId, memgraph::storage::v3::PropertyValue>> ConvertPropertyMap(
+    std::vector<std::pair<PropertyId, Value>> &properties) {
```

Comment:
I think the `properties` parameter should be also rvalue, to clearly express that the content of the vector will be moved.